File: app/src/main/java/com/chuongvd/app/signal/data/source/remote/datasource/UserRemoteDataSource.java
package com.chuongvd.app.signal.data.source.remote.datasource;

import com.chuongvd.app.signal.data.source.remote.request.LoginRequest;
import com.chuongvd.app.signal.data.source.remote.service.RequestHelper;
import com.chuongvd.app.signal.data.source.remote.service.rx.SchedulerUtils;
import com.chuongvd.app.signal.data.entity.UserEntity;
import io.reactivex.Observable;

public class UserRemoteDataSource implements IUserRemoteDataSource {
    private static UserRemoteDataSource sInstance;

    private UserRemoteDataSource() {

    }

    public static UserRemoteDataSource getInstance() {
        if (sInstance == null) {
            sInstance = new UserRemoteDataSource();
        }
        return sInstance;
    }

    @Override
    public Observable<UserEntity> login(LoginRequest loginRequest) {
        return RequestHelper.getRequest()
                .login(loginRequest)
                .flatMap(SchedulerUtils.convertData())
                .compose(SchedulerUtils.applyObservableAsync());
    }
}

File: app/src/main/java/com/chuongvd/app/signal/data/source/remote/datasource/UserRemoteDataSource.kt
package com.chuongvd.app.signal.data.source.remote.datasource

import com.chuongvd.app.signal.data.entity.UserEntity
import com.chuongvd.app.signal.data.source.remote.request.LoginRequest
import com.chuongvd.app.signal.data.source.remote.service.RequestHelper
import com.chuongvd.app.signal.data.source.remote.service.rx.SchedulerUtils
import io.reactivex.Observable

class UserRemoteDataSource private constructor() : IUserRemoteDataSource {

  override fun login(loginRequest: LoginRequest): Observable<UserEntity> {
    return RequestHelper.request
        .login(loginRequest)
        .flatMap(SchedulerUtils.convertData<UserEntity>())
        .compose(SchedulerUtils.applyObservableAsync<UserEntity>())
  }

  companion object {
    private var sInstance: UserRemoteDataSource? = null

    val instance: UserRemoteDataSource
      get() {
        if (sInstance == null) {
          sInstance = UserRemoteDataSource()
        }
        return this.sInstance!!
      }
  }
}
